### ISSUE TITLE
Unable to Create Entity with Custom Repository (Null Exception in FluidityEntityService.SaveEntity)

### ISSUE THREAD

**Issue by Nicholas-Westby** [mattbrailsford/umbraco-fluidity#94]
Here's the relevant portion of the stack trace from the Umbraco error log file:

```text
 2019-03-01 15:43:40,998 [P6720/D15/T11] ERROR Fluidity.Web.Api.FluidityApiController - Unhandled controller exception occurred
System.NullReferenceException: Object reference not set to an instance of an object.
   at Fluidity.Services.FluidityEntityService.SaveEntity(FluidityCollectionConfig collection, Object entity)
   at Fluidity.Web.Api.FluidityApiController.SaveEntity(FluidityEntityPostModel postModel)
   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ActionExecutor.<>c__DisplayClass10.<GetExecutor>b__9(Object instance, Object[] methodParameters)
   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ExecuteAsync(HttpControllerContext controllerContext, IDictionary`2 arguments, CancellationToken cancellationToken)
```

Here's my model:

```c#
public class SamplePerson
{
    public string Id { get; set; }
    public string FirstName { get; set; }
    public string LastName { get; set; }
}
```

Here's my custom repository:

```c#
using Fluidity;
using Fluidity.Data;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using Umbraco.Core.Models;

public class SampleRepo : FluidityRepository<SamplePerson, string>
{
    private static List<SamplePerson> People = new List<SamplePerson>()
    {
        new SamplePerson()
        {
            Id = "Bob",
            FirstName = "Bob",
            LastName = "Boberson"
        },
        new SamplePerson()
        {
            Id = "Jane",
            FirstName = "Jane",
            LastName = "Janerson"
        }
    };

    protected override void DeleteImpl(string id)
    {
        People = People.Where(x => x.Id != id).ToList();
    }

    protected override IEnumerable<SamplePerson> GetAllImpl()
    {
        return People;
    }

    protected override string GetIdImpl(SamplePerson entity)
    {
        return entity?.Id;
    }

    protected override SamplePerson GetImpl(string id)
    {
        return People.Where(x => x.Id == id).FirstOrDefault();
    }

    protected override PagedResult<SamplePerson> GetPagedImpl(int pageNumber, int pageSize,
        Expression<Func<SamplePerson, bool>> whereClause, Expression<Func<SamplePerson, object>> orderBy,
        SortDirection orderDirection)
    {
        var pageIndex = pageNumber - 1;
        var compiledWhere = whereClause == null
            ? new Func<SamplePerson, bool>(x => true)
            : whereClause.Compile();
        var compiledOrder = orderBy == null
            ? new Func<SamplePerson, object>(x => x.FirstName)
            : orderBy.Compile();
        var result = People
            .Where(x => compiledWhere(x))
            .OrderBy(x => compiledOrder(x))
            .Skip(pageIndex * pageSize)
            .Take(pageSize)
            .ToList();
        return new PagedResult<SamplePerson>(People.Count, pageNumber, pageSize)
        {
            Items = result
        };
    }

    protected override long GetTotalRecordCountImpl()
    {
        return People.Count;
    }

    protected override SamplePerson SaveImpl(SamplePerson entity)
    {
        if (string.IsNullOrWhiteSpace(entity.Id))
        {
            entity.Id = Guid.NewGuid().ToString("N");
        }
        People.Add(entity);
        return entity;
    }
}
```

And here's my bootstrap file:

```c#
using Fluidity.Configuration;

public class SampleBootstrap : FluidityConfigModule
{
    public override void Configure(FluidityConfig config)
    {
        config.AddSection("Fluidity Sample", sampleConfig =>
        {
            sampleConfig.SetTree("Sample Tree", treeConfig =>
            {
                treeConfig.AddCollection<SamplePerson>(x => x.Id, "Person", "People", "A collection of people", collConfig =>
                {
                    collConfig.SetRepositoryType<SampleRepo>();
                    collConfig.SetNameProperty(y => y.FirstName);
                    collConfig.SetViewMode(Fluidity.FluidityViewMode.List);
                    collConfig.ShowOnDashboard();
                    collConfig.ListView(listView =>
                    {
                        //listView.AddField(y => y.FirstName);
                        listView.AddField(y => y.LastName);
                    });

                    collConfig.Editor(editorConfig =>
                    {
                        editorConfig.AddTab("About", tab =>
                        {
                            //tab.AddField(y => y.FirstName).MakeRequired().SetDescription("The first name of this individual.");
                            tab.AddField(y => y.LastName).MakeRequired();
                        });
                        editorConfig.AddTab("Meta", tab =>
                        {
                            tab.AddField(y => y.Id).MakeReadOnly();
                        });
                    });
                });
            });
        });
    }
}
```

It mostly works. When I try to save a new entity, it bombs out, like this:

![save-error](https://user-images.githubusercontent.com/5933222/53672961-1fc4a580-3c42-11e9-8a9b-c301211a4af9.png)

Not really sure what the issue could be. Maybe I've missed some config or maybe it's a Fluidity bug relating to custom repositories?

More info:

* I set a breakpoint in every method in my `SampleRepo` class. None of them are called when I try to save an entity.
* I commented out a couple lines because I thought Fluidity might not like the name property also being an editable property. Didn't seem to help.
* Umbraco 7.13.2
* Fluidity 1.0.3

**Comment by Nicholas-Westby** [comment]
Looks like the problem is on this line:

![save-entity](https://user-images.githubusercontent.com/5933222/53674168-45a17880-3c49-11e9-83e6-d16abcea780e.png)

The portion that says `entity.GetPropertyValue(collection.IdProperty)` returns null, which makes sense, since this is a new entity and a new entity can have a null ID.

The `.Equals(` is then called on this null instance. Since `.Equals` is not an extension method, I imagine this is why it's bombing out (because you can't call methods on null instances).

My guess is that this is not usually an issue because you typically work with non-nullable value types (e.g., integer) for your ID types. I'm using a string as my ID in this case. So I imagine the fix is to ensure this works with nullable types by accounting for the possibility that the ID may be null.

**Comment by Nicholas-Westby** [comment]
Until the pull request gets merged, I created a weird workaround. I'll share it here for others in the same predicament.

The model class:

```c#
public class SamplePerson
{
    public SampleStringIdentifier Id { get; set; }
    public string FirstName { get; set; }
    public string LastName { get; set; }
}
```

Repo class:

```c#
using Fluidity;
using Fluidity.Data;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using Umbraco.Core.Models;

public class SampleRepo : FluidityRepository<SamplePerson, SampleStringIdentifier>
{
    private static List<SamplePerson> People = new List<SamplePerson>()
    {
        new SamplePerson()
        {
            Id = new SampleStringIdentifier("Bob"),
            FirstName = "Bob",
            LastName = "Boberson"
        },
        new SamplePerson()
        {
            Id = new SampleStringIdentifier("Jane"),
            FirstName = "Jane",
            LastName = "Janerson"
        }
    };

    protected override void DeleteImpl(SampleStringIdentifier id)
    {
        People = People.Where(x => x.Id.Id != id.Id).ToList();
    }

    protected override IEnumerable<SamplePerson> GetAllImpl()
    {
        return People;
    }

    protected override SampleStringIdentifier GetIdImpl(SamplePerson entity)
    {
        return entity.Id;
    }

    protected override SamplePerson GetImpl(SampleStringIdentifier id)
    {
        return People.Where(x => x.Id.Id == id.Id).FirstOrDefault();
    }

    protected override PagedResult<SamplePerson> GetPagedImpl(int pageNumber, int pageSize,
        Expression<Func<SamplePerson, bool>> whereClause, Expression<Func<SamplePerson, object>> orderBy,
        SortDirection orderDirection)
    {
        var pageIndex = pageNumber - 1;
        var compiledWhere = whereClause == null
            ? new Func<SamplePerson, bool>(x => true)
            : whereClause.Compile();
        var compiledOrder = orderBy == null
            ? new Func<SamplePerson, object>(x => x.FirstName)
            : orderBy.Compile();
        var result = People
            .Where(x => compiledWhere(x))
            .OrderBy(x => compiledOrder(x))
            .Skip(pageIndex * pageSize)
            .Take(pageSize)
            .ToList();
        return new PagedResult<SamplePerson>(People.Count, pageNumber, pageSize)
        {
            Items = result
        };
    }

    protected override long GetTotalRecordCountImpl()
    {
        return People.Count;
    }

    protected override SamplePerson SaveImpl(SamplePerson entity)
    {
        if (string.IsNullOrWhiteSpace(entity.Id.Id))
        {
            entity.Id = new SampleStringIdentifier(Guid.NewGuid().ToString("N"));
            People.Add(entity);
        }
        else
        {
            People = People.Select(x => x.Id.Id == entity.Id.Id ? entity : x).ToList();
        }
        return entity;
    }
}
```

Bootstrap file:

```c#
using Fluidity.Configuration;

public class SampleBootstrap : FluidityConfigModule
{
    public override void Configure(FluidityConfig config)
    {
        config.AddSection("Fluidity Sample", sampleConfig =>
        {
            sampleConfig.SetTree("Sample Tree", treeConfig =>
            {
                treeConfig.AddCollection<SamplePerson>(x => x.Id, "Person", "People", "A collection of people", collConfig =>
                {
                    collConfig.SetRepositoryType<SampleRepo>();
                    collConfig.SetNameProperty(y => y.FirstName);
                    collConfig.SetViewMode(Fluidity.FluidityViewMode.List);
                    collConfig.ShowOnDashboard();
                    collConfig.ListView(listView =>
                    {
                        //listView.AddField(y => y.FirstName);
                        listView.AddField(y => y.LastName);
                    });

                    collConfig.Editor(editorConfig =>
                    {
                        editorConfig.AddTab("About", tab =>
                        {
                            //tab.AddField(y => y.FirstName).MakeRequired().SetDescription("The first name of this individual.");
                            tab.AddField(y => y.LastName).MakeRequired();
                        });
                        editorConfig.AddTab("Meta", tab =>
                        {
                            tab.AddField(y => y.Id).MakeReadOnly();
                        });
                    });
                });
            });
        });
    }
}
```

New identifier struct:

```c#
[TypeConverter(typeof(SampleStringIdentifierConverter))]
[Serializable]
public struct SampleStringIdentifier
{
    public string Id { get; set; }
    public SampleStringIdentifier(string id)
    {
        Id = id;
    }
    public override string ToString()
    {
        return Id ?? string.Empty;
    }
}
```

Type converter for identifier struct:

```c#
using System;
using System.ComponentModel;
using System.Globalization;

public class SampleStringIdentifierConverter : TypeConverter
{
    public override bool CanConvertFrom(ITypeDescriptorContext context, Type sourceType)
    {
        return sourceType == typeof(string);
    }
    public override object ConvertFrom(ITypeDescriptorContext context, CultureInfo culture, object value)
    {
        if (value is string)
        {
            if (string.IsNullOrWhiteSpace(value as string))
            {
                return default(SampleStringIdentifier);
            }
            else
            {
                return new SampleStringIdentifier(value as string);
            }
        }
        return base.ConvertFrom(context, culture, value);
    }
}
```

Here's what I had to do:

* Create a struct since, unlike strings, structs are not nullable. Use this as the type on my ID field.
* That bombed out since Fludity couldn't convert from string to the struct type. So I created a converter.
* That bombed out because Fluidity didn't like an instance of the struct with a null string ID. Used empty string instead of null.
* That bombed out because an empty string got interpreted as a valid non-default ID and that ID didn't match any instances. Modified converter to treat an empty string as null so the struct gets created as a default instance.

Pretty wonky. Looking forward to the next release.

**Comment by mattbrailsford** [comment]
Should be fixed in 690f30a where I've swapped it to 

    var isNew = Equals(entity.GetPropertyValue(collection.IdProperty), collection.IdProperty.Type.GetDefaultValue());
